substrate/client/rpc-spec-v2/src/archive/archive_storage.rs

```diff
@@ -127,3 +140,738 @@ where
 		ArchiveStorageResult::ok(storage_results, discarded_items)
 	}
 }
+
+/// Parse hex-encoded string parameter as raw bytes.
+///
+/// If the parsing fails, returns an error propagated to the RPC method.
+pub fn parse_hex_param(param: String) -> Result<Vec<u8>, ArchiveError> {
+	// Methods can accept empty parameters.
+	if param.is_empty() {
+		return Ok(Default::default())
+	}
+
+	array_bytes::hex2bytes(&param).map_err(|_| ArchiveError::InvalidParam(param))
+}
+
+#[derive(Debug, PartialEq, Clone)]
+pub struct DiffDetails {
+	key: StorageKey,
+	return_type: ArchiveStorageDiffType,
+	child_trie_key: Option<ChildInfo>,
+	child_trie_key_string: Option<String>,
+}
+
+/// The type of storage query.
+#[derive(Debug, PartialEq, Clone, Copy)]
+enum FetchStorageType {
+	/// Only fetch the value.
+	Value,
+	/// Only fetch the hash.
+	Hash,
+	/// Fetch both the value and the hash.
+	Both,
+}
+
+/// The return value of the `fetch_storage` method.
+#[derive(Debug, PartialEq, Clone)]
+enum FetchedStorage {
+	/// Storage value under a key.
+	Value(StorageResult),
+	/// Storage hash under a key.
+	Hash(StorageResult),
+	/// Both storage value and hash under a key.
+	Both { value: StorageResult, hash: StorageResult },
+}
+
+pub struct ArchiveStorageDiff<Client, Block, BE> {
+	client: Storage<Client, Block, BE>,
+}
+
+impl<Client, Block, BE> ArchiveStorageDiff<Client, Block, BE> {
+	pub fn new(client: Arc<Client>) -> Self {
+		Self { client: Storage::new(client) }
+	}
+}
+
+impl<Client, Block, BE> ArchiveStorageDiff<Client, Block, BE>
+where
+	Block: BlockT + 'static,
+	BE: Backend<Block> + 'static,
+	Client: StorageProvider<Block, BE> + Send + Sync + 'static,
+{
+	/// Fetch the storage from the given key.
+	fn fetch_storage(
+		&self,
+		hash: Block::Hash,
+		key: StorageKey,
+		maybe_child_trie: Option<ChildInfo>,
+		ty: FetchStorageType,
+	) -> Result<Option<FetchedStorage>, String> {
+		match ty {
+			FetchStorageType::Value => {
+				let result = self.client.query_value(hash, &key, maybe_child_trie.as_ref())?;
+
+				Ok(result.map(FetchedStorage::Value))
+			},
+
+			FetchStorageType::Hash => {
+				let result = self.client.query_hash(hash, &key, maybe_child_trie.as_ref())?;
+
+				Ok(result.map(FetchedStorage::Hash))
+			},
+
+			FetchStorageType::Both => {
+				let value = self.client.query_value(hash, &key, maybe_child_trie.as_ref())?;
+				let Some(value) = value else {
+					return Ok(None);
+				};
+
+				let hash = self.client.query_hash(hash, &key, maybe_child_trie.as_ref())?;
+				let Some(hash) = hash else {
+					return Ok(None);
+				};
```

Comment:
```suggestion
				let Some(hash) = self.client.query_hash(hash, &key, maybe_child_trie.as_ref())? else {
				    return Ok(None);
				};
```

Total nit! I'd leave it be though if that makes the line too long and cargo fmt makes it uglier :D

Reply:
Have applied this for both hash and value, fmt looks good thanks 🙏